### PR TITLE
variable substitution added

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

File: Tasks/AzureRmWebAppDeployment/VariableSubstituter.ps1 [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        [string][Parameter(mandatory=$true)] 
+        $ConfigFileRegex
+    )
+
+    # Load modules
+    . $PSScriptRoot/XMLSubstituter.ps1 -Force
+
+    # Defined supported MIME type
+    $ApplicationXmlMimeType = "application/xml"
+    $TextXmlMimeType = "application/xml"
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/VariableSubstituter.ps1]: is this text/xml?

File: Tasks/AzureRmWebAppDeployment/VariableSubstituter.ps1 [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    $ConfigExtension = ".config"
+
+    # Find all configuration files
+    $webConfigFiles = Get-ChildItem -Path $WebAppFolderPath -Filter $ConfigFileRegex -Recurse -ErrorAction SilentlyContinue -Force
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/VariableSubstituter.ps1]: Find-VstsFiles also gives this functionality, may be it has some additional checks and balances. Do you think it is better to use that?

File: Tasks/AzureRmWebAppDeployment/VariableSubstituter.ps1 [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    if( $webConfigFiles.Count -eq 0)
+    {
+        Write-Verbose "File not found at location : $WebAppFolderPath with pattern $ConfigFileRegex"
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/VariableSubstituter.ps1]: Should this be a warning? As user said variable substitution but in actual that didn't happen

File: Tasks/AzureRmWebAppDeployment/VariableSubstituter.ps1 [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            Write-Verbose "Unable to detect MIME type for file with $($file.Extension) extension"
+            Continue
+        }
+
+        # Compare mime type of file and invoke corresponding variable substituter
+        if( $fileMimeType -eq $ApplicationXmlMimeType -or $fileMimeType -eq $TextXmlMimeType )
+        {
+            
+            $tags = @("appSettings","connectionStrings","configSections")
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/VariableSubstituter.ps1]: add spaces after comma

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            
+            $taskContextValueForAttribute = Get-VstsTaskVariable -Name $childNodeAttribute.Name 
+            if( $taskContextValueForAttribute ){
+                $childNode.SetAttribute($childNodeAttribute.Name,$taskContextValueForAttribute)
+            }
+        }
+        
+        # Check for key value pair of attribute
+        $valueForKeyAttribute = $childNode.GetAttribute("key")
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: remove unnecessary white spaces wherever applicable.

File: Tasks/AzureRmWebAppDeployment/task.loc.json [hunk 6]
```diff
@@ -146,7 +146,7 @@
     },
     {
       "name": "XdtFilesRoot",
-      "type": "string",
+      "type": "filePath",
       "label": "ms-resource:loc.input.label.XdtFilesRoot",
       "defaultValue": "",
       "groupName": "configtransforms",
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/task.loc.json]: This node remove please remove

**Comment by mvvsubbu** [comment]
Looks fine .. Keep your changes in your branch until my PR is done :)

File: Tasks/AzureRmWebAppDeployment/CompressionUtility.ps1 [hunk 7]
```diff
@@ -1,5 +1,4 @@
-﻿. $PSScriptRoot/Utility.ps1
-
+﻿
 function UnzipWebDeployPkg
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/CompressionUtility.ps1]: As discussed with Subbu, the temp path that you are using to extract, may not be a good idea. There are chances you will hit the max file path size limit when you extract zip with deep hierarchy.

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ps1 [hunk 8]
```diff
@@ -60,20 +62,29 @@ try{
...
+                if($env:RELEASE_ENVIRONMENTNAME.config)
+                {
+                    FindAndApplyTransformation -baseFile $configFile -tranformFile "web.$env:RELEASE_ENVIRONMENTNAME.config" -xdtFilesRoot $XdtFilesRoot
+                }
             }
-        }
+         }
+         if( $VariableSubstitution -eq "true" )
+         {
+            Substitute-Variables -WebAppFolderPath $unzippedPath -ConfigFileRegex "web.*config"
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ps1]: The spec says we should support any *.config. We are passing web.*config. Are we on same page?

File: Tasks/AzureRmWebAppDeployment/LegacyUtils/VariableSubstituter-Legacy.ps1 [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    # Load modules
+    . $PSScriptRoot/XMLSubstituter-Legacy.ps1 -Force
+
+    # Defined supported MIME type
+    $ApplicationXmlMimeType = "application/xml"
+    $TextXmlMimeType = "text/xml"
+    $ConfigExtension = ".config"
+
+    # Find all configuration files
+    $webConfigFiles = Get-ChildItem -Path $WebAppFolderPath -Filter $ConfigFileRegex -Recurse -ErrorAction SilentlyContinue -Force
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/LegacyUtils/VariableSubstituter-Legacy.ps1]: The xml substitution logic seem to find files using our lib helper and not trying to find all these Mime types. Why it has to be two different logic for variable substitution and xml transformation?

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ps1 [hunk 10]
```diff
@@ -60,20 +62,29 @@ try{
...
-        foreach ($configFile in $webconfigFiles) {
-            FindAndApplyTransformation -baseFile $configFile -tranformFile "web.release.config" -xdtFilesRoot $XdtFilesRoot
-            if($env:RELEASE_ENVIRONMENTNAME.config)
-            {
-                FindAndApplyTransformation -baseFile $configFile -tranformFile "web.$env:RELEASE_ENVIRONMENTNAME.config" -xdtFilesRoot $XdtFilesRoot
+        
+        if( $XmlTransformation -eq "true" )
+        {
+            # Search for all the web.config files
+            $webconfigFiles = Find-VstsFiles -LegacyPattern "$unzippedPath\**\web.config" -IncludeFiles
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ps1]: finding files and iterating on them are all moved to respective helper in other helpers (Zip, Variable substitution). Can we have same for xml transformation as well?

File: Tasks/AzureRmWebAppDeployment/LegacyUtils/XMLSubstituter-Legacy.ps1 [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        foreach( $childNodeAttribute in $childNodeAttributes )
+        {
+            $taskContextValueForAttribute = Get-TaskVariable $distributedTaskContext $childNodeAttribute.Name 
+            if( $taskContextValueForAttribute ){
+                $childNode.SetAttribute($childNodeAttribute.Name,$taskContextValueForAttribute)
+            }
+        }
+        
+        # Check for key value pair of attribute
+        $valueForKeyAttribute = $childNode.GetAttribute("key")
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/LegacyUtils/XMLSubstituter-Legacy.ps1]: Can we sync-up in person. I have few queries on scenarios.

File: Tasks/AzureRmWebAppDeployment/Strings/resources.resjson/en-US/resources.resjson [hunk 12]
```diff
@@ -31,8 +31,8 @@
   "loc.input.help.AdditionalArguments": "Additional Web Deploy arguments that will be applied when deploying the Azure Web App like,-disableLink:AppPoolExtension -disableLink:ContentExtension.",
   "loc.input.label.XmlTransformation": "Transformation",
   "loc.input.help.XmlTransformation": "The config transforms will be run for \\*.Release.config and \\*.&#060;EnvironmentName&#062;.config on all the web.config files.",
-  "loc.input.label.XdtFilesRoot": "Xdt Transformation Files Path",
-  "loc.input.help.XdtFilesRoot": "Root folder for searching for \\*.Release.config and \\*.&#060;EnvironmentName&#062;.config recursively.",
+  "loc.input.label.VariableSubstitution": "Variable Substitution",
+  "loc.input.help.VariableSubstitution": "Variables defined in the Defination will be automatically substitued in the appSettings,applicationSettings and connectionStrings section of the config files. If the same variable are defined in the Defination and Environment, then the En variables will be supersede the defination variables. Variable substitution is run after config Transformations.",
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/Strings/resources.resjson/en-US/resources.resjson]: The help about Definition and Environment is relevant only in Release. This help string may not be relevant in build.

**Comment by Ajay-MS** [comment]
closing it as there is already a PR for master master branch for variable substitution

**Pull request closed without merging**